### ISSUE TITLE
Ecrire crashes on start when using Wayland as the compositor.

### ISSUE THREAD

**Issue by ApostolosB** [Obsidian-StudiosInc/ecrire#2]
Backtrace:
```
#0  0xb69d4b10 in XGetSelectionOwner () from /usr/lib/libX11.so.6
[Current thread is 1 (Thread 0xb5658e40 (LWP 3109))]
(gdb) bt
#0  0xb69d4b10 in XGetSelectionOwner () at /usr/lib/libX11.so.6
#1  0xb71b8250 in ecore_x_selection_owner_get (atom=0) at lib/ecore_x/ecore_x_selection.c:712
#2  0x0804c723 in _start ()
(gdb)
```

**Comment by wltjr** [comment]
Great thanks! Mentioned on IRC just restating  here for anyone who comes across such. I am not able to run Wayland yet myself due to lack of multi-head/display support and/or switching displays. As such I do not really have means to run or test under Wayland short of VM etc. Its low on my priorities list, but is something I will address. Contributions are welcome.

Otherwise when I get time I will look at the docs and see if I can adapt the code for Wayland. I also need to update cmake so it does not enable such when Wayland is used. Presently activates ecore_x if found and likely causes the issues under wayland.

Even without the issues, build without support for ecore x, still need to code the same features ecore x uses, under wayland.

Two issue;

1. Fix CMakeLists.txt, make ecore x or wayland conditional (prevents crash)
2. Code same function under wayland that is present under ecore x (function the same under either)

**Comment by wltjr** [comment]
I have a potential temporary fix for this in git. I added a cmake option wayland, ```cmake -Dwayland=true .``` to enable support for wayland. Which will use wayland over x. By default it will use ecore x, this will switch to ecore wl2; which is presently not used by ecrire. This should fix the crash issue. The ecore x code will not be included/compiled when wayland is enabled in ecrire cmake build.

If you could test to confirm if this fixes the crash at least. I know the keyboard shortcuts will be disabled for most actions. I will have to see if that code is really ecore x specific or generic input code. This should fix the crash at least, and make ecrire run under wayland, hopefully....

Thanks for the report!

**Comment by NuLogicSystems** [comment]
Raster just disabled the x specific code so it works on wayland, no need to have a option in cmake
https://git.enlightenment.org/apps/ecrire.git/commit/?id=20983dc2a2c737620c7aaf8f45d47d5e84e71334

**Comment by wltjr** [comment]
I see that, and sadly that will make merging my work back into that repo more difficult. I was about to address this the proper way switching that code to elm_cnp.

**Comment by wltjr** [comment]
If Raster had looked at my code. Setting -Dwayland as stated above. Would have the same effect as commenting out that code. Except it would retain the function for people not under wayland. Commenting it out disables it for all. Though no work has been done in that repo. It does make me concerned about my work going back in that repo. If another is just going to make any modifications they feel like and potentially cause me more time and work to address....

**Comment by NuLogicSystems** [comment]
If what raster did doesn't exist in your code when you make your commit, then it shouldn't make a difference. Just git clone ecrire from git.enlightenment.org, then overwrite their files with yours when copying your code into your local git, then commit and push the changes, this should revert rasters commit in the process.

**Comment by wltjr** [comment]
It does not work that way. He modified code that my modifications are based on. That breaks the ability to merge and creates a conflict. Try it yourself. Checkout that repo. Then in a branch checkout mine. Try to merge mine into that one and you will get a conflict.

**Comment by NuLogicSystems** [comment]
I didn't say anything about merge, copy from your repo and paste into the cloned repo.

**Comment by wltjr** [comment]
That is not the proper way. I have made hundreds of commits since then.

**Comment by NuLogicSystems** [comment]
I know it's not the proper way, but it should work. How about this, since it lloks like only one commit why not `git reset HEAD`, or if it is more than one commit then `git rebase master~20983dc2a2c737620c7aaf8f45d47d5e84e71334 master~<whatever the last commit was that you forked the code at> master`.

**Comment by wltjr** [comment]
That is one way but it is not good to do on public repos. It will require a force push and anyone pulling will have issues. To fix will just require a commit to reverse the other commit. Then I can merge/rebase on top of that. It just was pointless to even touch that repo. If anything should have pulled in my changes. But it was a hack fix not proper. My way is a more correct temporary fix till I can replace the code with generic. As I had been looking into. This goes back a while. If you look at this ticket [T5476](https://phab.enlightenment.org/T5476). They closed a duplicate one of that ticket where Tom was looking into the same years ago. All this is known and being worked on.

**Comment by NuLogicSystems** [comment]
Well, however you and/or raster get it worked out. ;)

**Comment by wltjr** [comment]
Yes its minor, its not really a big deal. Just wish he would have looked first, or asked before proceeding.That is really the only concern. Once my work is merged in and/or I have commit access. Others just jumping in and making blind changes is not really ideal. Its more a warning than anything. But all things can be fixed and this is minor. Just a good example that work flow need be coordinated a bit. Likely not practical for ecrire, but many times for various things on community projects, patches are submitted for review before inclusion. That is probably an extreme for ecrire. Just looking to see if anything is active already and/or checking with primary maintainers is all :)

**Comment by wltjr** [comment]
Resolved with recent commit. At least any crash relating to ecore x code. May still crash under wayland for other reasons. Crashing on save is known. Anything else would need a new ticket. Closing this as resolved.

**Comment by ApostolosB** [comment]
just compiled from git and still see:

```
(gdb) bt
#0  0xb69dbb10 in XGetSelectionOwner () at /usr/lib/libX11.so.6
#1  0xb71c94d2 in ecore_x_selection_owner_get (atom=0) at lib/ecore_x/ecore_x_selection.c:712
#2  0x0804ca1f in _start ()
(gdb)

```

**Comment by wltjr** [comment]
Is it still crashing or just linking to that stuff? Seems like cmake is linking in lots of unused and unwanted stuff.

**Comment by NuLogicSystems** [comment]
I just tried to create a new build and now have a cmake error after the commits on the 15th.

**Comment by ApostolosB** [comment]
Crashes when i start it. Compiled it with j1, it compiled fine, tried to run it and got the above.

FWIW i also tried to launch it from a tty. Crashes there also.

**Comment by NuLogicSystems** [comment]
Can you post some debuging info with gdb and or ddd?

**Comment by wltjr** [comment]
Reopening issue till confirmed resolved.

The above seems to be output from running in GDB not a normal crash output etc. Though it seems odd. It is not using any of that stuff anymore. Nor should it be linking to it unless something else is causing that. The ecrire binary is linked to a bunch of stuff not used. Something is pulling that in and maybe EFL itself. I will have to look into why that is. Seems rather excessive.

Also part of these problems come from having support for both ecore x and wayland. I think if your system is wayland only. No ecore x modules, etc. Then it cannot link to that stuff and cause any runtime issues. Though many will likely have both so do need to come up with a fix.

As for the TTY issue. I need to code up something so it bails there. I doubt I will bother trying to fire up a x/wayland server to run ecrire from TTY. Assuming you mean a regular non graphical terminal without a graphic session running. Not sure if any EFL stuff supports starting from TTY and not within a graphical env. Though I could be mistaken, and you maybe starting it from a terminal. Though I think by TTY you mean non GUI terminal.

**Comment by wltjr** [comment]
What branch are you pulling from? Make sure it is master. I merged my branch into master a bit ago. I need to delete my old branch, wltjr.

**Comment by ApostolosB** [comment]
OK. Now it works. At least renders on WL and from a TTY.

Closing.

**Comment by wltjr** [comment]
Great, maybe was pulling from wrong branch. I had the default set to my old I deleted. That may have caused a problem. Either way good to know thanks!
